gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java

```diff
@@ -0,0 +1,281 @@
+
+        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+
+        Request request = requestBuilder
+                .setPreserveAuthorizationOnRedirect(true)
+                .setFollowRedirects(false)
+                .build();
+
+        FluentFuture<ProxyResponse> future = executeHttp(request);
+
+        if (request.getUri().getPath().startsWith(V1_STATEMENT_PATH) && request.getMethod().equals(HttpMethod.POST)) {
+            future = future.transform(response -> recordBackendForQueryId(request, response), executor);
+        }
+
+        setupAsyncResponse(
+                asyncResponse,
+                future.transform(response -> buildResponse(response, oauth2GatewayCookie), executor)
+                        .catching(ProxyException.class, e -> handleProxyException(request, e), directExecutor()));
+    }
+
+    private ImmutableList<NewCookie> getOAuth2GatewayCookie(URI remoteUri, HttpServletRequest servletRequest)
+    {
+        if (cookiesEnabled) {
+            if (remoteUri.getPath().startsWith(OAuth2GatewayCookie.OAUTH2_PATH)
+                    && !(servletRequest.getCookies() != null
+                    && Arrays.stream(servletRequest.getCookies()).anyMatch(c -> c.getName().equals(OAuth2GatewayCookie.NAME)))) {
+                GatewayCookie oauth2Cookie = new OAuth2GatewayCookie(getRemoteTarget(remoteUri));
+                return ImmutableList.of(oauth2Cookie.toNewCookie());
+            }
+            else if (servletRequest.getCookies() != null) {
+                return Arrays.stream(servletRequest.getCookies())
+                        .filter(c -> c.getName().startsWith(GatewayCookie.PREFIX))
+                        .map(GatewayCookie::fromCookie)
+                        .filter(c -> !c.isValid() || c.matchesDeletePath(remoteUri.getPath()))
+                        .map(GatewayCookie::toNewCookie)
+                        .map(newCookie -> new NewCookie.Builder(newCookie).value("delete").maxAge(0).build())
+                        .collect(toImmutableList());
+            }
+        }
+        return ImmutableList.of();
+    }
+
+    private static String getRemoteTarget(URI remoteUri)
+    {
+        return format("%s://%s", remoteUri.getScheme(), remoteUri.getAuthority());
+    }
+
+    private Response buildResponse(ProxyResponse response, ImmutableList<NewCookie> cookie)
+    {
+        Response.ResponseBuilder builder = Response.status(response.statusCode()).entity(response.body());
+        response.headers().forEach((headerName, value) -> builder.header(headerName.toString(), value));
+        cookie.forEach(builder::cookie);
+        return builder.build();
+    }
+
+    private void setupAsyncResponse(AsyncResponse asyncResponse, ListenableFuture<Response> future)
+    {
+        bindAsyncResponse(asyncResponse, future, executor)
+                .withTimeout(ASYNC_TIMEOUT, () -> Response
+                        .status(BAD_GATEWAY)
```

Comment:
trino-client [will retry](https://github.com/trinodb/trino/blob/cfe9ac9f73232c89990bba1354767c038e5997d6/client/trino-client/src/main/java/io/trino/client/HttpStatusCodes.java#L24) on a `BAD_GATEWAY` response code until its own request timeout is exceeded. If I follow the code correctly, it looks like the JDBC driver defaults to a request timeout of [2 minutes](https://github.com/trinodb/trino/blob/b575d7af6ef66efdae441a45c1427c951d209dcd/client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java#L95), so with `ASYNC_TIMEOUT` = 2 minutes, the JDBC driver will not retry. I think we should reduce ASYNC_TIMEOUT, or make it configurable, to give the client one or more chances to retry.

Reply:
A retry from client may or may not be a desirable behavior. It could cause a duplicated submission of a query.
I'll make `ASYNC_TIMEOUT` configurable and let the admin choose what they want.
I'd like to keep the default to 2m, which is the same as [trino-proxy](https://github.com/trinodb/trino/blob/0c0f86c97694f231ac61438a1a1a3d7e82668a61/service/trino-proxy/src/main/java/io/trino/proxy/ProxyResource.java#L95).